### PR TITLE
implemented the registration service and module tests.

### PR DESCRIPTION
**Pull request by mhorianin**

### CODE REVIEW

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
+package core.basesyntax.service;
+
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private RegistrationService registrationService = new RegistrationServiceImpl();
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:9]: Try to init in Before all method

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+import core.basesyntax.model.User;
+import org.junit.jupiter.api.Test;
+
+class RegistrationServiceImplTest {
+    private RegistrationService registrationService = new RegistrationServiceImpl();
+
+    @Test
+    void registerUserNull_NotOk() {
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:12]: Use this pattern  for naming for all tests
image

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+        User bobPasswordNull = new User();
+        bobPasswordNull.setAge(19);
+        bobPasswordNull.setLogin("bobist");
+        bobPasswordNull.setPassword("");
+        bobPasswordNull.setId(52698L);
+
+        assertThrows(InvalidDataException.class, () -> {
+            registrationService.register(bobPasswordNull);
+        }, "An exception should be thrown");
+
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:46]: You test have naming nullUser, but I see 3 kinds of user and no one user is null!!! divide to few tests

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        User alice = new User();
+        alice.setAge(15);
+        alice.setLogin("aliceFox");
+        alice.setPassword("hhjf562bv");
+        alice.setId(36898L);
+
+        assertThrows(InvalidDataException.class, () -> {
+            registrationService.register(alice);
+        }, "An exception should be thrown");
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:69]: Place assert in the end of method

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 5]
```diff
@@ -5,10 +5,27 @@
 import core.basesyntax.model.User;
 
 public class RegistrationServiceImpl implements RegistrationService {
+    private static final int MAX_AGE = 18;
+    private static final int MIN_NUMBER_CHARACTERS = 6;
     private final StorageDao storageDao = new StorageDaoImpl();
```
[the remark this answers is not in the document]
Comment (mhorianin) [reply src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:10]: I don't understand in which constructor

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 6]
```diff
@@ -5,10 +5,36 @@
...
     public User register(User user) {
-        return null;
+        if (user == null) {
+            throw new RegistrationException("User is invalid. Please enter the data");
+        }
+        if (user.getAge() == null) {
+            throw new RegistrationException("The age is invalid");
+        }
+        if (user.getAge() < MAX_AGE) {
+            throw new RegistrationException("The user must be at least 18 years old");
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:21]: replace hardcode with max age constant

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 7]
```diff
@@ -5,10 +5,36 @@
...
+        }
+        if (storageDao.get(user.getLogin()) != null) {
+            throw new RegistrationException("A user with this login already exists");
+        }
+        if (user.getLogin() == null) {
+            throw new RegistrationException("The login is invalid");
+        }
+        if (user.getLogin().length() < MIN_NUMBER_CHARACTERS) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:29]: include limit in the message

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java [hunk 8]
```diff
@@ -5,10 +5,36 @@
...
+            throw new RegistrationException("The login is short");
+        }
+        if (user.getPassword() == null) {
+            throw new RegistrationException("The password is invalid");
+        }
+        if (user.getPassword().length() < MIN_NUMBER_CHARACTERS) {
```
Comment (okuzan) [inline src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:35]: include limit in the message

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+    private static StorageDao storageDao;
+
+    @BeforeAll
+    static void beforeAll() {
+        registrationService = new RegistrationServiceImpl();
+        storageDao = new StorageDaoImpl();
+    }
+
+    @Test
+    void register_validUser_Ok() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:23]: use lowercaser after underscore
```suggestion
    void register_validUser_ok() {

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+        john.setPassword("jo");
+        john.setId(1259L);
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(john);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_lengthLogin_notOk() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:180]:
```suggestion
    void register_shortLogin_notOk() {
```
друга частина для опису того, що даємо на вхід тестовому методу

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+        User bob = new User();
+        bob.setAge(19);
+        bob.setLogin("bob");
+        bob.setPassword("hhjfklk");
+        bob.setId(52698L);
+
+        User john = new User();
+        john.setAge(19);
+        john.setLogin("john");
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:189]: edge case for login length is 5-6, both should be tested (6 is tested in _ok method)

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+            registrationService.register(bob);
+        }, "An exception should be thrown");
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(alice);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_negativeAge_notOk() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:114]: one test user is enough, you can include it in the invalid age test

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,201 @@
...
+        bob.setAge(10);
+        bob.setLogin("bobist");
+        bob.setPassword("hhjfbchbv");
+        bob.setId(52698L);
+
+        User alice = new User();
+        alice.setAge(15);
+        alice.setLogin("aliceFox");
+        alice.setPassword("hhjf562bv");
+        alice.setId(36898L);
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:102]: test only one user with pre-edge case 17

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 14]
```diff
@@ -0,0 +1,201 @@
...
+        bob.setPassword("hhjfbchbv");
+        bob.setId(52698L);
+        registrationService.register(bob);
+
+        User alice = new User();
+        alice.setAge(18);
+        alice.setLogin("aliceFox");
+        alice.setPassword("hhjf562bv");
+        alice.setId(36898L);
+        registrationService.register(alice);
```
Comment (okuzan) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:36]: you should assert the object returned by the register method is correct too

**Review by okuzan — CHANGES_REQUESTED** [review]
Consider creating a user field, and updating only its relevant properties in each test, that would greatly reduce the amount of code

File: src/main/java/core/basesyntax/service/RegistrationServiceImpl.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kozhukhovsky) [reply src/main/java/core/basesyntax/service/RegistrationServiceImpl.java:10]: need create constructor

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+        john.setId(1259L);
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(john);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_shortLogin_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:157]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+        registrationService.register(john);
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(bob);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_shortPassword_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:137]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+        }, "An exception should be thrown");
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(bobAgeNegative);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_invalidLogin_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:120]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+        bob.setPassword(null);
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(bob);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_invalidAge_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:90]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+        bob.setLogin(null);
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(bob);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_nullPassword_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:80]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 20]
```diff
@@ -0,0 +1,174 @@
...
+        bob.setAge(null);
+
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(bob);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_nullLogin_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:70]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+    @Test
+    void register_nullUser_notOk() {
+        assertThrows(RegistrationException.class, () -> {
+            registrationService.register(null);
+        }, "An exception should be thrown");
+    }
+
+    @Test
+    void register_nullAge_notOk() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:60]:
```suggestion

```

File: src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java [hunk 22]
```diff
@@ -0,0 +1,174 @@
...
+        bob = new User();
+        bob.setAge(20);
+        bob.setLogin("bobiik");
+        bob.setPassword("hhjfbchbv");
+        bob.setId(52698L);
+    }
+
+    @Test
+    void register_validUser_ok() {
+
```
Comment (kozhukhovsky) [inline src/test/java/core/basesyntax/service/RegistrationServiceImplTest.java:35]:
```suggestion

```

**Review by kozhukhovsky — APPROVED** [review]
